### ISSUE TITLE
Not properly serving cached files and making new copies instead.

### ISSUE THREAD

**Issue by axipher** [steamcache/steamcache#10]
So I'm trying this out with two simple games, Hammerwatch and Screencheat.  I download them initially and it shows in the watchlog that each chunk is being pulled form remote in red text.  Then I uninstall the game in Steam and attempt to reinstall it.  Looking at the log it will match the first 1-5 chunks as local in green, then proceed to go red for all the remaining chunks and redownload them all.  I can see the "depot" folder growing in size and opening up a chunks directory, I can see that the same chunk is being stored multiple times with an extra set of numbers at the end:

"\\MILLENNIUM\steam_cache\depot\301971\chunk\00c4437516e6554d80f4e030dff2cf602e6fc3f6.0000009361"
"\\MILLENNIUM\steam_cache\depot\301971\chunk\00c4437516e6554d80f4e030dff2cf602e6fc3f6.0000012640"
"\\MILLENNIUM\steam_cache\depot\301971\chunk\00c4437516e6554d80f4e030dff2cf602e6fc3f6.0000002422"
"\\MILLENNIUM\steam_cache\depot\301971\chunk\00c4437516e6554d80f4e030dff2cf602e6fc3f6.0000005500"

**Comment by axipher** [comment]
![image](https://cloud.githubusercontent.com/assets/25443249/22468639/0ebc8a20-e797-11e6-8fa2-fafa963fc37a.png)

**Comment by JasonRivers** [comment]
I will look in to this in the next couple of weeks, I'm currently waiting for internet to be installed. but as a side note, be careful running the watchlog, We used to show this by default but it turns out to be a huge CPU hog on large networks. It also uses steamdb to get the name of the depot, which could lead to an IP ban on steamdb, I'll patch this so that it doesn't do the requests to steamdb to save the IP Ban. It will still show red/green/yellow depending on the content.

**Comment by axipher** [comment]
Okay, so just don't run the Watchlog all the time then, I only have one machine configured to use this right now until it works, then it will be for a house with 8 gaming PC's in it.  If it makes any difference, the guide I used to set it all up was from Ars Technica: https://arstechnica.com/gaming/2017/01/building-a-local-steam-caching-server-to-ease-the-bandwidth-blues/

When you get a chance to update, will it simply be a case of deleting the contents of the current depot folder, then remove the steamcache docker itself and re-add it so it pulls down the latest one?

**Comment by JasonRivers** [comment]
You can just run docker pull steamcache/steamcache:latest and then stop your current container and run another one, you shouldn't have to remove the depot folder at all.

**Comment by webberist** [comment]
I updated csgo and watched with wire shark and none of my URLs had a dot in
them. FYI, My isp's content server is one of the ones with ?etime=... in
the URL.

Unfortunately I haven't purchased the two games you tested. Can you try it
with something free like TF2?
I suspect it's a problem specific to either certain game titles or certain
content servers.

**Comment by handruin** [comment]
I'm writing to confirm I've seen the same issue as @axipher while testing a single game title through Steam.  I tested the game Guacamelee! Gold Edition by downloading the game, uninstalling it, and the downloading the game again.  I would see a mixture of red and green remote/local from the watchlog.sh script as only some of the chunks would download from cache.  I also observed the Steam client's network usage graph would speed up for a few seconds and slow down as it appeared to change from local cache to internet download.

I tested this with another game The Witcher 3 and I noticed similar behavior although it was more difficult to tell how frequently it would download chucks from remote vs local because the download size was much larger.  I made sure to pull the latest steamcache docker image from this evening while.

![watcher_client](https://cloud.githubusercontent.com/assets/8650004/22624097/813f64ae-eb3f-11e6-9985-626fb63cb31b.jpg)

**Comment by ilumos** [comment]
All the red cache miss URLs there have `&h=4f47...` - check if those chunk IDs have a different `h` value when first downloaded?

There's a remote chance that `h` is a hash of the chunk indicating it has changed, which requires it to be re-cached. 

Or it could be that the steam client changing the value of `h` is causing Nginx to flag as a cache miss, though that doesn't seem right as the variable `$uri` (URL without query string) is used in the `try_files` directive, rather than `$request_uri` (URL with query string):

```
location /depot/ {
	try_files $uri @mirror;
}
```
Hope this helps narrow down the cause

**Comment by kixelated** [comment]
Why is steamcache using `try_files` instead of `proxy_cache` like generic? I'm kind of worried about writing user-defined paths to disk. `proxy_cache` also hashes the uri which avoids a few potential issues.

**Comment by Azothyran** [comment]
Yeah I'm having this problem as well. Installed GTA V from Steam last night, uninstalled it, then reinstalled it and nearly half the files are missing or not being read from the cache. There have been no updates for the game since I warmed the cache.

![steamcache missing chunks](https://cloud.githubusercontent.com/assets/26119600/23469635/beb4d4b2-fe69-11e6-9dee-998effda1074.PNG)

Any kind of testing I can provide to help rectify this I'm down for.

**Comment by GotenXiao** [comment]
@kixelated The steamcache image was originally built to allow the data to be shared between people's caches (i.e. for LAN party prep use) - it uses `proxy_store` in the `@mirror` location instead of `proxy_cache`, which doesn't save the result files with the query string parameters, but does pass them on to the CDN when the cache requests a missing file. This should produce a closer image of what's on Steam's CDNs than the generic image, and is slightly easier to purge specific content from (as well as easier to share between cache instances via sneakernet).

@Azothyran I can confirm the presence of chunk 2a06f8067c7820825d51b3887656c8f9a278f7ab in my local cache, for that depot ID, so I'm not entirely sure what's going on with your config. Have you bind mounted an external directory for depot storage by passing `-v` when running the container?

**Comment by Azothyran** [comment]
@GotenXiao I passed `/srv/steamcache/data/` on the physical disk to `/data/cache/` in the docker container.

My specific setup if it's necessary uses Ubuntu Server as a Hyper-V guest, and I have a 2TB disk attached directly to the server to use as an OS drive and for steamcache storage.

**Comment by kixelated** [comment]
@GotenXiao Yeah, I understand it's easier to just store Valve's URI format to disk. The alternative `proxy_cache` stores based on the md5 of the URI so it's just as shareable but you're right, purging is more annoying. I'm just paranoid about file names based on unsanitized user input but that's off-topic.

My only suggestion is to try `proxy_store $uri` instead of `proxy_store on`.  Also, try `proxy_temp_path /data/tmp` to avoid copying files between disks which could be causing the numbers at the end.

And finally for a shameless plug, but my fork combines steamcache and generic and shouldn't have this issue. https://github.com/kixelated/steamcache

**Comment by ekhoury2000** [comment]
i Used a windows share to create a repository, i can see in watchlog red entries but windows directory remains empty :( any suggestions?

The command I ran was
$ sudo docker run --name steamcache --restart=always -d -v /srv/steamcache/data:/data/cache -p 80:80 steamcache/steamcache:latest

i think the issue is there, should i create directories on my ubuntu server like /data/cache?
my windows share is //192.168.1.1/cache /srv/steamcache/data
thanks

**Comment by Azothyran** [comment]
@ekhoury2000 I'm assuming you're using fstab to automount the directory on boot? If so, I'll bet you neglected to give your Linux machine write permissions to the SMB share. Assuming it's an anonymous share that does NOT require a login, an example follows.

`//192.168.1.1/cache /srv/steamcache/data/ cifs guest,iocharset=utf8,file_mode=0777,dir_mode=0777,noexec,noperm 0 0`

That `file_mode` and `dir_mode` delegates who can write to it; without it, you'll run into an issue where writing to that share fails.

**Comment by ekhoury2000** [comment]
@Azothyran Thank you so much, i had used the ./smdcredentials file to logon to the shared dir, seemed not to work but i did as u said and it worked perfectly fine, the respository is now being filled up.

Thanks a lot 👍

**Comment by ekhoury2000** [comment]
After Testing it, i tried downloaing a game from steam, download is stuck and says queued, and if i try to force by launching the game, i get the below error:

An error occurred while updating counter strike: corrupt download:
valve503.steampipe.steamcontent.com.depot.chunk..........

is it possible this is 1 game issue? or this might affect all my download at a certain time?

Thanks a lot

**Comment by Azothyran** [comment]
After shutting down and removing the docker images, flushing my cache directory, and pulling the two docker images in sequence by these commands, it seems to be working perfectly now as far as I can see.

For the steamcache image:
`sudo docker run --name steamcache -d -v /srv/steamcache/data:/data/cache -p 80:80 steamcache/steamcache:latest --restart=always`

Then after that's pulled and running, I ran this for the steamcache-dns image:
`sudo docker run --name steamcache-dns -p 192.168.15.3:53:53/udp -e STEAMCACHE_IP=192.168.15.3 steamcache/steamcache-dns:latest --restart always`

I tried the same method of testing as before; uninstalling GTA V and reinstalling it overnight, then attempt a reinstall while watching the cache the next day, and all the files seem to be properly stored and served as far as I can tell. Only strange thing I noticed was that after setting up the steamcache-dns image, the process stopped at `Starting bind9` and then never moves on (I just closed the terminal window after 10 minutes to make sure it wasn't supposed to move on by itself) so I'd assume there isn't an exit call that's made after bind9 starts...

At any rate, steamcache seems to be doing its thing now; I have no idea what the difference between my previous attempt and this one was that caused it to not work properly last time.

![steamcache working](https://cloud.githubusercontent.com/assets/26119600/24471630/58356ba4-1488-11e7-9992-6c4067de27e0.jpg)
